common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -59,114 +67,124 @@ func (s *localStoreCertProvider) GetSettings() *config.GroupTLS {
 }
 
 func (s *localStoreCertProvider) FetchServerCertificate() (*tls.Certificate, error) {
+
+	if s.tlsSettings == nil {
+		return nil, nil
+	}
 	return s.FetchCertificate(&s.serverCert, s.tlsSettings.Server.CertFile, s.tlsSettings.Server.CertData,
 		s.tlsSettings.Server.KeyFile, s.tlsSettings.Server.KeyData)
 }
 
 func (s *localStoreCertProvider) FetchClientCAs() (*x509.CertPool, error) {
-	if len(s.tlsSettings.Server.ClientCAFiles) == 0 && len(s.tlsSettings.Server.ClientCAData) == 0 {
-		return nil, nil
-	}
 
-	s.RLock()
-	if s.clientCAs != nil {
-		defer s.RUnlock()
-		return s.clientCAs, nil
-	}
-
-	s.RUnlock()
-	s.Lock()
-	defer s.Unlock()
-
-	if s.clientCAs != nil {
-		return s.clientCAs, nil
-	}
-
-	clientCaPoolFromFiles, err := buildCAPoolFromFiles(s.tlsSettings.Server.ClientCAFiles)
-	if err != nil {
-		return nil, err
+	if s.tlsSettings == nil {
+		return nil, nil
 	}
+	return s.fetchCAs(
+		s.tlsSettings.Server.ClientCAFiles,
+		s.tlsSettings.Server.ClientCAData,
+		&s.clientCAs,
+		&s.clientCACerts,
+		"cannot specify both clientCAFiles and clientCAData properties")
+}
 
-	clientCaPoolFromData, err := buildCAPoolFromData(s.tlsSettings.Server.ClientCAData)
-	if err != nil {
-		return nil, err
-	}
+func (s *localStoreCertProvider) FetchServerRootCAsForClient(isWorker bool) (*x509.CertPool, error) {
 
-	if clientCaPoolFromFiles != nil && clientCaPoolFromData != nil {
-		return nil, errors.New("cannot specify both clientCAFiles and clientCAData properties")
+	clientSettings := s.getClientTLSSettings(isWorker)
+	if clientSettings == nil {
+		return nil, nil
 	}
+	rootCAFiles := clientSettings.RootCAFiles
+	rootCAData := clientSettings.RootCAData
 
-	if clientCaPoolFromData != nil {
-		s.clientCAs = clientCaPoolFromData
+	var cached **x509.CertPool
+	var certs *[]*x509.Certificate
+	if isWorker {
+		cached = &s.serverCAsWorker
+		certs = &s.serverCACertsWorker
 	} else {
-		s.clientCAs = clientCaPoolFromFiles
+		cached = &s.serverCAs
+		certs = &s.serverCACerts
 	}
 
-	return s.clientCAs, nil
+	return s.fetchCAs(
+		rootCAFiles,
+		rootCAData,
+		cached,
+		certs,
+		"cannot specify both rootCAFiles and rootCAData properties")
 }
 
-func (s *localStoreCertProvider) FetchServerRootCAsForClient(isWorker bool) (*x509.CertPool, error) {
-	clientSettings := s.getClientTLSSettings(isWorker)
-	rootCAFiles := clientSettings.RootCAFiles
-	rootCAData := clientSettings.RootCAData
-
-	if len(rootCAFiles) == 0 && len(rootCAData) == 0 {
+func (s *localStoreCertProvider) fetchCAs(
+	files []string,
+	data []string,
+	cached **x509.CertPool,
+	certs *[]*x509.Certificate,
```

Comment:
Also I would not modify parameters but add another return value.

Reply:
in order to modify a passed in slice, you do have to pass a pointer to it.
That being said, I agree its easier if we return stuff rather than setting it via parameters